### PR TITLE
drivers: i3c: npcx: i3c target driver

### PR DESCRIPTION
**Pull request by alvsun**
1. add buffer mode callback api for DMA.
2. add npcx i3c targt driver.

### CODE REVIEW

File: drivers/i3c/i3c_npcx.c [hunk 1]
```diff
@@ -1835,57 +1924,359 @@ static int npcx_i3c_ibi_disable(const struct device *dev, struct i3c_device_desc
...
+		return -EINVAL;
+	}
+
+	if ((is_i3c_start | is_i3c_stop) == 0) {
+		LOG_ERR("%s: not the end of xfer, is_start: %d, is_stop:%d", __func__, is_i3c_start,
+			is_i3c_stop);
+		return -EINVAL;
+	}
+
+	if (get_oper_state(dev) == NPCX_I3C_OP_STATE_WR) { /* Write request */
```
Comment (fabiobaltieri) [inline drivers/i3c/i3c_npcx.c:2182]: could you invert this condition, handle the other case with the existing code and a goto and unindent the rest?

File: drivers/i3c/i3c_npcx.c [hunk 2]
```diff
@@ -2057,63 +2459,401 @@ static int npcx_i3c_cntlr_init(const struct device *dev)
...
+	}
+
+	/* Set Provisional ID */
+	pid = config_target->pid;
+
+	/* PID[47:33] MIPI manufacturer ID */
+	SET_FIELD(inst->VENDORID, NPCX_I3C_VENDORID_VID, (uint32_t)GET_PID_VENDOR_ID(pid));
+
+	/* PID[32] Verdor fixed value(0) or random value(1) */
+	if (config_target->pid_random == true) {
```
Comment (fabiobaltieri) [inline drivers/i3c/i3c_npcx.c:2516]: `	if (config_target->pid_random) {`

pid_random is a bool

File: drivers/i3c/i3c_npcx.c [hunk 3]
```diff
@@ -2057,63 +2459,401 @@ static int npcx_i3c_cntlr_init(const struct device *dev)
...
+		LOG_ERR("Get APB4 source clock fail %d", ret);
+		return -EINVAL;
+	}
+
+	bamatch = get_bus_available_match_val(apb4_rate);
+	LOG_DBG("BAMATCH: %d", bamatch);
+	SET_FIELD(inst->CONFIG, NPCX_I3C_CONFIG_BAMATCH, bamatch);
+
+	/* Enable secondary controller capabillity if supported */
+	if (config->secondary_mode == true) {
```
Comment (fabiobaltieri) [inline drivers/i3c/i3c_npcx.c:2505]: ``if (config->secondary_mode) {

**Review by fabiobaltieri — COMMENTED** [review]
hey it looks like second commit has more than just the "add target driver", would it make sense to break it down to smaller changes? it's hard to follow as it is now

File: dts/bindings/i3c/nuvoton,npcx-i3c.yaml [hunk 4]
```diff
@@ -50,3 +50,55 @@ properties:
...
+      Bit[7:4] module id.
+      Bit[3:0] port id.
+
+  target-mode-en:
+    type: boolean
+    description: Initialized as controller or target mode.
+
+  secondary:
+    type: boolean
+    description: Initialized as a secondary master.
```
Comment (XenuIsWatching) [inline dts/bindings/i3c/nuvoton,npcx-i3c.yaml:67]: I don't understand why you need both of these. It either boots as a controller (primary controller) or a target (secondary controller)

File: dts/arm/nuvoton/npcx/npcx4.dtsi [hunk 5]
```diff
@@ -388,6 +388,8 @@
 			status = "disabled";
 			#address-cells = <3>;
 			#size-cells = <0>;
+
```
Comment (XenuIsWatching) [inline dts/arm/nuvoton/npcx/npcx4.dtsi:391]: remove extra line

File: dts/arm/nuvoton/npcx/npcx4.dtsi [hunk 6]
```diff
@@ -413,6 +415,8 @@
...
```
Comment (XenuIsWatching) [inline dts/arm/nuvoton/npcx/npcx4.dtsi:418]: remove extra line

File: dts/arm/nuvoton/npcx/npcx4.dtsi [hunk 7]
```diff
@@ -438,6 +442,8 @@
...
```
Comment (XenuIsWatching) [inline dts/arm/nuvoton/npcx/npcx4.dtsi:445]: remove extra line

File: drivers/i3c/i3c_npcx.c [hunk 8]
```diff
@@ -2144,41 +2884,55 @@ static int npcx_i3c_init(const struct device *dev)
...
 	if (ret != 0) {
 		LOG_ERR("Addr slots init fail %d", ret);
 		return ret;
 	}
 
-	ctrl_config->is_secondary = false; /* Currently can only act as primary controller. */
-	ctrl_config->supported_hdr = I3C_MSG_HDR_DDR; /* HDR-DDR mode is supported. */
-	ctrl_config->scl.i3c = config->clocks.i3c_pp_scl_hz; /* Set I3C frequency */
+	/* Set controller default configuration */
+	config_cntlr->is_secondary = false; /* Currently can only act as primary controller. */
```
Comment (XenuIsWatching) [inline drivers/i3c/i3c_npcx.c:2901]: it makes more sense to have that dts prop to only define if it is a secondary controller. If it's a secondary controller, then it should initialize as a target, if it is a primary controller, then it should initialize as a controller.

The only way to have to 'switch' at runtime would be to do a controller handoff rather than calling the configure function to have it initialize as a target.

**Review by XenuIsWatching — CHANGES_REQUESTED** [review]

File: dts/bindings/i3c/nuvoton,npcx-i3c.yaml (hunk 4 above)
Comment (alvsun) [reply dts/bindings/i3c/nuvoton,npcx-i3c.yaml:67]: The initial reason for doing this was to provide the option for the target device with secondary controller capability when it boots up as a target.

It means it can initialize as a pure target or as a target with a secondary controller.

image

File: drivers/i3c/i3c_npcx.c [hunk 9]
```diff
@@ -289,6 +348,29 @@ static inline void npcx_i3c_interrupt_enable(struct i3c_reg *inst, uint32_t mask
...
+	struct i3c_reg *inst = config->base;
+
+	/* Disable the target interrupt events */
+	inst->INTCLR = inst->INTSET;
+
+	/* Clear the target interrupt status */
+	inst->STATUS = inst->STATUS;
+
+	/* Enable the target interrupt events */
+	if (enable == true) {
```
[the remark this answers is not in the document]
Comment (alvsun) [reply drivers/i3c/i3c_npcx.c]: Done

File: drivers/i3c/i3c_npcx.c [hunk 10]
```diff
@@ -1835,57 +1924,359 @@ static int npcx_i3c_ibi_disable(const struct device *dev, struct i3c_device_desc
...
+	int index;
+
+	/* the request or the payload were not specific */
+	if ((request == NULL) || ((request->payload_len) && (request->payload == NULL))) {
+		return -EINVAL;
+	}
+
+	/* the I3C was not in target mode or the bus is in HDR mode now */
+	if ((!IS_BIT_SET(inst->CONFIG, NPCX_I3C_CONFIG_TGTENA)) ||
+	    (IS_BIT_SET(inst->STATUS, NPCX_I3C_STATUS_STHDR))) {
```
[the remark this answers is not in the document]
Comment (alvsun) [reply drivers/i3c/i3c_npcx.c]: Done

File: drivers/i3c/i3c_npcx.c [hunk 11]
```diff
@@ -1835,57 +1924,359 @@ static int npcx_i3c_ibi_disable(const struct device *dev, struct i3c_device_desc
...
+	mdma_inst->MDMA_CTL0 &= ~BIT(NPCX_MDMA_CTL_MDMAEN);
+	mdma_inst->MDMA_CTL0 &= ~BIT(NPCX_MDMA_CTL_TC); /* W0C */
+	mdma_inst->MDMA_CTL0 &= ~BIT(NPCX_MDMA_CTL_SIEN);
+	SET_FIELD(i3c_inst->DMACTRL, NPCX_I3C_DMACTRL_DMAFB, MDMA_DMAFB_DISABLE);
+
+	/* Ignore DA and detect all START and STOP */
+	i3c_inst->CONFIG &= ~BIT(NPCX_I3C_CONFIG_MATCHSS);
 
+	/* Flush the tx and rx FIFO */
+	i3c_inst->DATACTRL |= (BIT(NPCX_I3C_DATACTRL_FLUSHTB) | BIT(NPCX_I3C_DATACTRL_FLUSHFB));
```
[the remark this answers is not in the document]
Comment (alvsun) [reply drivers/i3c/i3c_npcx.c]: Done

File: drivers/i3c/i3c_npcx.c [hunk 12]
```diff
@@ -1835,57 +1924,359 @@ static int npcx_i3c_ibi_disable(const struct device *dev, struct i3c_device_desc
...
+	mdma_inst->MDMA_CTL1 &= ~BIT(NPCX_MDMA_CTL_MDMAEN);
+	mdma_inst->MDMA_CTL1 &= ~BIT(NPCX_MDMA_CTL_TC); /* W0C */
+	mdma_inst->MDMA_CTL1 &= ~BIT(NPCX_MDMA_CTL_SIEN);
+	SET_FIELD(i3c_inst->DMACTRL, NPCX_I3C_DMACTRL_DMATB, MDMA_DMATB_DISABLE);
+
+	/* Ignore DA and detect all START and STOP */
+	i3c_inst->CONFIG &= ~BIT(NPCX_I3C_CONFIG_MATCHSS);
+
+	/* Flush the tx and rx FIFO */
+	i3c_inst->DATACTRL |= (BIT(NPCX_I3C_DATACTRL_FLUSHTB) | BIT(NPCX_I3C_DATACTRL_FLUSHFB));
```
[the remark this answers is not in the document]
Comment (alvsun) [reply drivers/i3c/i3c_npcx.c]: Done

File: drivers/i3c/i3c_npcx.c (hunk 2 above)
Comment (alvsun) [reply drivers/i3c/i3c_npcx.c]: Done

File: drivers/i3c/i3c_npcx.c (hunk 3 above)
Comment (alvsun) [reply drivers/i3c/i3c_npcx.c:2511]: Done

File: dts/bindings/i3c/nuvoton,npcx-i3c.yaml (hunk 4 above)
Comment (alvsun) [reply dts/bindings/i3c/nuvoton,npcx-i3c.yaml]: Done.
Please refer https://github.com/alvsun/zephyr/blob/e1622037e214d7dd93a4f3246b2544e42ae4afa0/drivers/i3c/i3c_npcx.c#L2911 #

File: dts/arm/nuvoton/npcx/npcx4.dtsi (hunk 5 above)
Comment (alvsun) [reply dts/arm/nuvoton/npcx/npcx4.dtsi]: Done

File: dts/arm/nuvoton/npcx/npcx4.dtsi (hunk 7 above)
Comment (alvsun) [reply dts/arm/nuvoton/npcx/npcx4.dtsi]: Done

File: drivers/i3c/i3c_npcx.c (hunk 8 above)
Comment (alvsun) [reply drivers/i3c/i3c_npcx.c]: Done

File: drivers/i3c/i3c_npcx.c [hunk 13]
```diff
@@ -2144,41 +2893,53 @@ static int npcx_i3c_init(const struct device *dev)
...
+	config_cntlr->is_secondary = config->secondary_mode;
+	config_cntlr->supported_hdr = I3C_MSG_HDR_DDR;        /* HDR-DDR mode is supported. */
+	config_cntlr->scl.i3c = config->clocks.i3c_pp_scl_hz; /* Set I3C frequency */
 
-	ret = npcx_i3c_configure(dev, I3C_CONFIG_CONTROLLER, ctrl_config);
+	/* Initial the device as controller or target */
+	config_type = (config_target->enable | config_cntlr->is_secondary) ? I3C_CONFIG_TARGET
+									   : I3C_CONFIG_CONTROLLER;
+	config_ptr = (config_type == I3C_CONFIG_CONTROLLER) ? (void *)config_cntlr
+							    : (void *)config_target;
```
Comment (XenuIsWatching) [inline drivers/i3c/i3c_npcx.c:2918]: This still does not make sense to me. If someone has `is_secondary` set to true and `config_target->enable` set to false. Then this would not be a valid startup condition for I3C according to the spec because as a secondary controller you must startup as a target.
I'd suggest remove this `target_mode_en` prop and only use `is_secondary`. If a device is to be seen as a "Target-Only" device, then you are to set the proper values in the BCR so another controller won't try to do a handoff to the device.

File: drivers/i3c/i3c_npcx.c [hunk 14]
```diff
@@ -2232,16 +3005,28 @@ static const struct i3c_driver_api npcx_i3c_driver_api = {
...
 		IF_ENABLED(CONFIG_I3C_NPCX_DMA, (                                                  \
 			.mdma_clk_subsys = NPCX_DT_CLK_CFG_ITEM_BY_IDX(id, 2),                     \
 		))                                                                                 \
-		IF_ENABLED(CONFIG_I3C_NPCX_DMA, (                                                  \
+				     IF_ENABLED(CONFIG_I3C_NPCX_DMA, (                             \
 			.mdma_base = (struct mdma_reg *)DT_INST_REG_ADDR_BY_IDX(id, 1),            \
-		))                                                                                 \
+		)) };                                                                              \
+	static struct npcx_i3c_data npcx_i3c_data_##id = {                                         \
+		.config_target.enable = DT_INST_PROP_OR(id, target_mode_en, false),                \
```
Comment (XenuIsWatching) [inline drivers/i3c/i3c_npcx.c:3019]: change to 
```c
.config_target.enable = DT_INST_PROP_OR(id, secondary, false), \
```

File: dts/bindings/i3c/nuvoton,npcx-i3c.yaml [hunk 15]
```diff
@@ -50,3 +50,57 @@ properties:
...
+    type: int
+    description: |
+      Instance ID of the device, used to specify port number.
+      Bit[7:4] module id.
+      Bit[3:0] port id.
+
+  target-mode-en:
+    type: boolean
+    description: |
+      0: controller mode, 1: target mode.
```
Comment (XenuIsWatching) [inline dts/bindings/i3c/nuvoton,npcx-i3c.yaml:65]: remove

**Review by XenuIsWatching — CHANGES_REQUESTED** [review]

File: drivers/i3c/i3c_npcx.c (hunk 13 above)
Comment (XenuIsWatching) [inline drivers/i3c/i3c_npcx.c:2918]: This still doesn't make sense to me. What if config target is true and `is_secondary` is false. If it is primary controller, then it should always boot as a controller... but this will skip initializing it as a controller

Comment (alvsun) [reply drivers/i3c/i3c_npcx.c:2918]: A little bit confused here.
The I3C module of the NPCX4 supports both controller and target functionality.
Before booting up the device, it can be configured to one of the following modes.
I have planned the overlay configuration as follows to correspond to the different modes:
|    | target-mode-en   |secondary  | config_type  (Result) |
|  ----  | ----  | ----  | ---- |
| Primary Controller  | false |false |I3C_CONFIG_CONTROLLER |
| Secondary Controller  | false |true |I3C_CONFIG_TARGET |
| Ordinary Target (No controller capability)  | true |false |I3C_CONFIG_TARGET |

Not sure if I understand your meaning correctly.
So if `config_target = true `and `is_secondary = false`, it means the user would like to initialize the I3C device as an ordinary target at startup.

File: drivers/i3c/i3c_npcx.c [hunk 16]
```diff
@@ -2057,63 +2464,405 @@ static int npcx_i3c_cntlr_init(const struct device *dev)
...
+		SET_FIELD(inst->MCONFIG, NPCX_I3C_MCONFIG_CTRENA, MCONFIG_CTRENA_OFF);
+	}
+
+	/* Set Provisional ID */
+	pid = config_target->pid;
+
+	/* PID[47:33] MIPI manufacturer ID */
+	SET_FIELD(inst->VENDORID, NPCX_I3C_VENDORID_VID, (uint32_t)GET_PID_VENDOR_ID(pid));
+
+	/* PID[32] Verdor fixed value(0) or random value(1) */
```
Comment (XenuIsWatching) [inline drivers/i3c/i3c_npcx.c:2522]: typo `Verdor` should be `Vendor`

File: drivers/i3c/i3c_npcx.c [hunk 17]
```diff
@@ -2057,63 +2464,405 @@ static int npcx_i3c_cntlr_init(const struct device *dev)
...
+	struct i3c_config_target *config_target;
 
 	if (type == I3C_CONFIG_CONTROLLER) {
+		config_cntlr = config;
 		/*
 		 * Check for valid configuration parameters.
 		 * Currently, must be the primary controller.
 		 */
-		if ((cntlr_cfg->is_secondary) || (cntlr_cfg->scl.i3c == 0U)) {
+		if ((config_cntlr->is_secondary) || (config_cntlr->scl.i3c == 0U)) {
```
Comment (XenuIsWatching) [inline drivers/i3c/i3c_npcx.c:2567]: You should still be able to configure as a controller while a secondary controller.... although Handoff support isn't quite here yet

File: drivers/i3c/i3c_npcx.c [hunk 18]
```diff
@@ -2057,63 +2464,405 @@ static int npcx_i3c_cntlr_init(const struct device *dev)
...
+			LOG_ERR("Configure controller failed");
 			return -EINVAL;
 		}
 
 		/* Save requested config to dev */
-		(void)memcpy(&dev_data->common.ctrl_config, cntlr_cfg, sizeof(*cntlr_cfg));
+		(void)memcpy(&dev_data->common.ctrl_config, config_cntlr, sizeof(*config_cntlr));
 
-		/* Controller init */
 		return npcx_i3c_cntlr_init(dev);
```
Comment (XenuIsWatching) [inline drivers/i3c/i3c_npcx.c:2575]: I see that this is set within the controller init code.... but it is a controller capable device it should still be able to configure the controller configuration for itself.  I'd have to recommend this only be set once within the init code based on the `secondary` prop from the dts
```c
	/* Enable main controller mode */
	SET_FIELD(inst->MCONFIG, NPCX_I3C_MCONFIG_CTRENA, MCONFIG_CTRENA_ON);
```

Comment (alvsun) [reply drivers/i3c/i3c_npcx.c]: Done (39e1aea)
Move the control of the I3C device role to` npcx_i3c_cntlr_init()` and `npcx_i3c_target_init()`.
Initial `npcx_i3c_cntlr_init()` or `npcx_i3c_target_init()` only in `npcx_i3c_init()`.

File: drivers/i3c/i3c_npcx.c [hunk 19]
```diff
@@ -2144,41 +2921,51 @@ static int npcx_i3c_init(const struct device *dev)
...
+	/* Set controller default configuration */
+	config_cntlr->supported_hdr = I3C_MSG_HDR_DDR;        /* HDR-DDR mode is supported. */
+	config_cntlr->scl.i3c = config->clocks.i3c_pp_scl_hz; /* Set I3C frequency */
 
-	ret = npcx_i3c_configure(dev, I3C_CONFIG_CONTROLLER, ctrl_config);
+	/* Initial the device as controller or target */
+	config_type = (config_target->enable | config_cntlr->is_secondary) ? I3C_CONFIG_TARGET
+									   : I3C_CONFIG_CONTROLLER;
+	ret = (config_type == I3C_CONFIG_CONTROLLER) ? npcx_i3c_cntlr_init(dev)
+						     : npcx_i3c_target_init(dev);
```
Comment (XenuIsWatching) [inline drivers/i3c/i3c_npcx.c:2945]: I still do not think this will cover a valid case... but then again I do not believe handoff support is in this driver yet (or at least within the Zephyr i3c framework yet) but this will likely not work when it comes...

What I'd expect here is that there would be only a `is_secondary` prop within the dts, and you've already got the initial BCR within the device tree where bits[7:6] define if it is a controller capable target or if it is a target only device which could be read instead of having a redundant parameter of the `target-mode-en` when bits[7:6] of that BCR already defines this. If you want to make it less confusing, maybe rename `is_secondary` to be just `

image

I would also expect both `npcx_i3c_cntlr_init` and `npcx_i3c_target_init` to be called within here.
```c
if (I3C_BCR_DEVICE_ROLE(config_target->bcr) == I3C_BCR_DEVICE_ROLE_I3C_CONTROLLER_CAPABLE) {
	if (config_cntlr->is_secondary) {
		/* Secondary Controller, so boot as a target */
		cntlr_dev_en = MCONFIG_CTRENA_CAPABLE; /* Controller mode capable */
		inst->CONFIG |= BIT(NPCX_I3C_CONFIG_TGTENA); /* Target mode enable */
	} else {
		/* Primary Controller, so boot as a controller */
		cntlr_dev_en = MCONFIG_CTRENA_ON; /* Controller mode on */
	}
	npcx_i3c_cntlr_init(dev);
	npcx_i3c_target_init(dev);
} else {
	cntlr_dev_en = MCONFIG_CTRENA_OFF; /* Controller mode off */
	npcx_i3c_target_init(dev);
	inst->CONFIG |= BIT(NPCX_I3C_CONFIG_TGTENA); /* Target mode enable */
}

SET_FIELD(inst->MCONFIG, NPCX_I3C_MCONFIG_CTRENA, cntlr_dev_en);
```
_Along with he setting of the `MCONFIG_CTRENA_*` removed in the `npcx_i3c_cntlr_init` and `npcx_i3c_target_init`._

but then again.. maybe there's an understanding with the IP with how `MCONFIG_CTRENA_CAPABLE`, `MCONFIG_CTRENA_ON`, and `MCONFIG_CTRENA_OFF` I don't know about with how it is used. Is the datasheet public (or under NDA?). If so, where I can take a look at it?

File: drivers/i3c/i3c_npcx.c (hunk 16 above)
Comment (alvsun) [reply drivers/i3c/i3c_npcx.c]: Done

File: drivers/i3c/i3c_npcx.c (hunk 19 above)
Comment (alvsun) [reply drivers/i3c/i3c_npcx.c:2945]: Couple of questions about handoff I'd aligned with
1. If npcx4 bootup as controller mode, the i3c framwrok in zephyr does not support handoff yet.
2. If bootup as target mode, the npcx currently can do the handoff through npcx_i3c_target_ibi_raise(). 
   If handoff success, it'll swith to controller in target isr (Line 2845).
3. 
```c
if (I3C_BCR_DEVICE_ROLE(config_target->bcr) == I3C_BCR_DEVICE_ROLE_I3C_CONTROLLER_CAPABLE) {
	npcx_i3c_cntlr_init(dev);
	npcx_i3c_target_init(dev);
	if (config_cntlr->is_secondary) {
		/* Secondary Controller, so boot as a target */
		SET_FIELD(inst->MCONFIG, NPCX_I3C_MCONFIG_CTRENA, MCONFIG_CTRENA_CAPABLE);
		inst->CONFIG |= BIT(NPCX_I3C_CONFIG_TGTENA); /* Target mode enable */
	} else {
		/* Primary Controller, so boot as a controller */
		SET_FIELD(inst->MCONFIG, NPCX_I3C_MCONFIG_CTRENA, MCONFIG_CTRENA_ON);
	}
} else {
	npcx_i3c_target_init(dev);
	SET_FIELD(inst->MCONFIG, NPCX_I3C_MCONFIG_CTRENA, MCONFIG_CTRENA_OFF);
	inst->CONFIG |= BIT(NPCX_I3C_CONFIG_TGTENA); /* Target mode enable */
}
```

Comment (alvsun) [reply drivers/i3c/i3c_npcx.c:2945]: Thanks for your help. 
I'm still checking if the datasheet is public, and it might take a while.

> I still do not think this will cover a valid case... but then again I do not believe handoff support is in this driver yet (or at least within the Zephyr i3c framework yet) but this will likely not work when it comes...

Couple of questions about handoff I'd  aligned with, just in case I missed something during the implementation process.
1. For controller mode, the i3c framwrok in zephyr does not support handoff yet.
   (`I3C_IBI_CONTROLLER_ROLE_REQUEST` in `i3c_ibi_workq.c ` still in progress)
2. If bootup as target mode, the npcx currently can do the handoff through ibi_raise(). 
   If handoff success, it'll swith to controller in target isr.
    ```c
    static void npcx_i3c_target_isr(const struct device *dev)
    {
    
     .... 
	    /* Secondary controller (Controller register).
	     * Check I3C now bus controller.
	     * Disable target mode if target switch to controller mode success.
	     */
	    if (IS_BIT_SET(inst->MINTMASKED, NPCX_I3C_MINTMASKED_NOWCNTLR)) {
		    inst->MSTATUS = BIT(NPCX_I3C_MSTATUS_NOWCNTLR); /* W1C */
		    inst->CONFIG &= ~BIT(NPCX_I3C_CONFIG_TGTENA);   /* Disable target mode */
		    config_cntlr->is_secondary = false;             /* Switch to primary controller */
		    npcx_i3c_apply_cntlr_config(dev);
	    }
    
    ....
    }
    ```
    So, the current implementation does not support transferring bus control in controller mode but does support requesting bus control in target mode.

3. For the suggested init sequence with BCR, based on the code above, the target will init as controller once the controller request is success.
So I update the code as below. 
```c
if (I3C_BCR_DEVICE_ROLE(config_target->bcr) == I3C_BCR_DEVICE_ROLE_I3C_CONTROLLER_CAPABLE) {
       // npcx_i3c_apply_cntlr_config(dev);     /* Removed */
       // npcx_i3c_apply_target_config(dev);  /* Removed */  

	if (config_cntlr->is_secondary) {
                npcx_i3c_apply_target_config(dev);  /* Add Initial target here .... */
		SET_FIELD(inst->MCONFIG, NPCX_I3C_MCONFIG_CTRENA, MCONFIG_CTRENA_CAPABLE);
		inst->CONFIG |= BIT(NPCX_I3C_CONFIG_TGTENA);
	} else {
                npcx_i3c_apply_cntlr_config(dev); /* Add Initial controller here ....*/
		SET_FIELD(inst->MCONFIG, NPCX_I3C_MCONFIG_CTRENA, MCONFIG_CTRENA_ON);
	}
} else {
       ....
}
```

Comment (XenuIsWatching) [reply drivers/i3c/i3c_npcx.c]: I don't think you quite got the terminology right here (and yes it is confusing...)... but after the active controller does a handoff with GETACCCR to itself, it becomes the new "Active Controller". Keep in mind that Active Controller is different from the terms Primary and Secondary Controller. The Primary Controller is the "1st" Active Controller on the Bus. All other devices are targets (or controller-capable targets (secondary controllers). When the handoff happens, it is still called a "secondary controller", but it is now the "active controller" of which only one can exist on an I3C bus at a time. You shouldn't need these lines below in there when it comes.
```c
config_cntlr->is_secondary = false;             /* Switch to primary controller */
npcx_i3c_apply_cntlr_config(dev);
```

Anyways... I see you made the changes, so it LGTM now. Thanks!!

File: drivers/i3c/i3c_npcx.c [hunk 20]
```diff
@@ -2232,16 +3020,27 @@ static const struct i3c_driver_api npcx_i3c_driver_api = {
...
 			.mdma_base = (struct mdma_reg *)DT_INST_REG_ADDR_BY_IDX(id, 1),            \
-		))                                                                                 \
+		)) };                                                                              \
+	static struct npcx_i3c_data npcx_i3c_data_##id = {                                         \
+		.common.ctrl_config.is_secondary = DT_INST_PROP_OR(id, secondary, false),          \
+		.config_target.static_addr = DT_INST_PROP_OR(id, static_address, 0),               \
+		.config_target.pid = ((uint64_t)DT_INST_TGT_PID_PROP_OR(id, tgt_pid, 0) << 32) |   \
+				     DT_INST_TGT_PID_PROP_OR(id, tgt_pid, 1),                      \
+		.config_target.pid_random = DT_INST_TGT_PID_RAND_PROP_OR(id, tgt_pid, 0),          \
+		.config_target.bcr = DT_INST_PROP_OR(id, bcr, 0),                                  \
```
Comment (XenuIsWatching) [inline drivers/i3c/i3c_npcx.c:3038]: you have `bcr` set with required: true in the yaml, but yet this is with `DT_INST_PROP_OR` which implies it could not be set in the dts. Is this your intent?

File: drivers/i3c/i3c_npcx.c [hunk 21]
```diff
@@ -2232,16 +3020,27 @@ static const struct i3c_driver_api npcx_i3c_driver_api = {
...
 		.common.dev_list.num_i2c = ARRAY_SIZE(npcx_i3c_i2c_device_array_##id),             \
 		.pincfg = PINCTRL_DT_INST_DEV_CONFIG_GET(id),                                      \
+		.instance_id = DT_INST_PROP_OR(id, instance_id, 0),                                \
 		.clocks.i3c_pp_scl_hz = DT_INST_PROP_OR(id, i3c_scl_hz, 0),                        \
 		.clocks.i3c_od_scl_hz = DT_INST_PROP_OR(id, i3c_od_scl_hz, 0),                     \
 		IF_ENABLED(CONFIG_I3C_NPCX_DMA, (                                                  \
 			.mdma_clk_subsys = NPCX_DT_CLK_CFG_ITEM_BY_IDX(id, 2),                     \
 		))                                                                                 \
-		IF_ENABLED(CONFIG_I3C_NPCX_DMA, (                                                  \
+				     IF_ENABLED(CONFIG_I3C_NPCX_DMA, (                             \
```
Comment (XenuIsWatching) [inline drivers/i3c/i3c_npcx.c:3029]: nit: whitespace tab'ed over

Comment (alvsun) [reply drivers/i3c/i3c_npcx.c]: Done

File: drivers/i3c/i3c_npcx.c (hunk 20 above)
Comment (alvsun) [reply drivers/i3c/i3c_npcx.c]: Done.
Update `instance_id` and `bcr `  with `DT_INST_PROP`.

File: drivers/i3c/i3c_npcx.c (hunk 19 above)
Comment (alvsun) [reply drivers/i3c/i3c_npcx.c]: Done.
These two lines are removed.

**Review by XenuIsWatching — APPROVED** [review]

**Review by XenuIsWatching — APPROVED** [review]
it looks like there are some unrelated CI failures... you may need to rebase to get around it... but the PR now LGTM!

**Review by fabiobaltieri — APPROVED** [review]
